src/main/java/com/metamx/cache/CaffeineCache.java

```diff
@@ -45,15 +49,31 @@
 
   public static CaffeineCache create(final CaffeineCacheConfig config)
   {
-    Caffeine<Object, Object> builder = Caffeine.newBuilder().recordStats();
+    return create(config, null);
+  }
+
+  // Used in testing
+  public static CaffeineCache create(final CaffeineCacheConfig config, @Nullable final Executor executor)
+  {
+    Caffeine builder = Caffeine.newBuilder().recordStats();
     if (config.getExpiration() >= 0) {
       builder = builder
           .expireAfterAccess(config.getExpiration(), TimeUnit.MILLISECONDS);
     }
     if (config.getMaxSize() >= 0) {
       builder = builder
-          .maximumSize(config.getMaxSize())
-          .maximumWeight(config.getMaxSize());
+          .maximumWeight(config.getMaxSize())
+          .weigher(new Weigher<NamedKey, byte[]>()
```

Comment:
Well there is no reason to re-assign the builder since its returning `this`
for chaining. That works fine if you're then building with a lot of if-else
conditions. Otherwise you'd probably write it in a single shot like,

Cache<NamedKey, byte[]> cache = Caffeine.newBuilder()
.weigher((NamedKey key, byte[] val) -> key.namespace.length() + val.length)
.maximumWeight(maximumSizeInBytes)
.recordStats()
.build();

On Thu, Mar 31, 2016 at 12:40 PM, Charles Allen <notifications@github.com>
wrote:

> In src/main/java/com/metamx/cache/CaffeineCache.java
> <https://github.com/metamx/druid-cache-caffeine/pull/7#discussion_r58115152>
> :
>
> >      if (config.getExpiration() >= 0) {
> >        builder = builder
> >            .expireAfterAccess(config.getExpiration(), TimeUnit.MILLISECONDS);
> >      }
> >      if (config.getMaxSize() >= 0) {
> >        builder = builder
> > -          .maximumSize(config.getMaxSize())
> > -          .maximumWeight(config.getMaxSize());
> > +          .maximumWeight(config.getMaxSize())
> > +          .weigher(new Weigher<NamedKey, byte[]>()
>
> example:
>
>     Caffeine<Object, Object> builder = Caffeine.newBuilder().recordStats();
>    builder = builder.weigher((Cache.NamedKey key, byte[] val) -> key.namespace.length() + val.length);
>
> Java doesn't like that.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/metamx/druid-cache-caffeine/pull/7/files/00b3f0a3a5dec232e12bc135a6ccc27a738bc9b7#r58115152>
>